scipy/sparse/csgraph/_flow.pyx

```diff
@@ -668,3 +679,740 @@ cdef ITYPE_t[:] _dinic(
+                           result)
+        j = result[0]
+        s = result[1]
+        t = result[2]
+        _augment_flow(Wn, We, Wne_len,
+                      _residual_capacity(j, s, edge_sources,
+                                         edge_capacities,
+                                         edge_flow),
+                      edge_sources, edge_flow)
+        # Do nothing more if the entering edge is the same as the leaving edge.
+        if i != j:
+            if parent[t] != s:
+                # Ensure that s is the parent of t.
+                tmp = t
+                t = s
+                s = tmp
+            if _index(We, Wne_len[1], i) > _index(We, Wne_len[1], j):
+                # Ensure that q is in the subtree rooted at t.
+                tmp = q
+                q = p
+                p = tmp
+            _remove_edge(s, t, subtree_size,
+                         prev_vertex_dft,
+                         last_descendent_dft,
+                         next_vertex_dft,
+                         parent, parent_edge,
+                         n_verts)
+            _make_root(n_verts, q,
+                       subtree_size,
+                       last_descendent_dft,
+                       prev_vertex_dft,
+                       next_vertex_dft,
+                       parent, parent_edge,
+                       ancestors)
+            _add_edge(i, p, q,
+                      last_descendent_dft,
+                      next_vertex_dft,
+                      prev_vertex_dft,
+                      subtree_size,
+                      parent, parent_edge)
+            _update_potentials(i, p, q,
+                               edge_targets, edge_weights,
+                               last_descendent_dft,
+                               next_vertex_dft,
+                               vertex_potentials, subtree)
+        prev_ret_value = _find_entering_edges(n_non_zero_edges, edge_weights,
+                                              vertex_potentials, edge_flow,
+                                              edge_sources, edge_targets,
+                                              prev_ret_value, local_vars,
+                                              result)
+        itr += 1
+
+    result[3] = 1
+    for v in range(n_verts):
+        if edge_flow[v + n_non_zero_edges] != 0:
+            result[3] = 0
+            break
+
+    if result[3] == 1:
+        flow_cost = 0
+        flow_value = 0
+        for i in range(n_edges):
+            flow_cost += edge_weights[i]*edge_flow[i]
+            flow_value += edge_flow[i]
+        idx = 0
+        for i in range(n_edges):
+            if edge_flow[i] != 0 and edge_capacities[i] != 0:
+                _add_entry(edge_sources[i], edge_targets[i],
+                           edge_flow[i], idx, row, col, flow_data)
+                idx += 1
+        result[0] = flow_cost
+        result[1] = idx
+        result[2] = flow_value
+    return result
```

Comment:
When possible, it's better to allocate everything once using numpy to then be able to release the GIL explicitly, helping the Cython code to be mapped to efficient C code.

In your implementation, this looks possible (:+1:), so let's switch.

Note that the following restructuring:
 - is a draft and has not been tested.
 - removes `edge_flow_final`, which is not being used.
 - makes use of some syntactic sugar for variables' definition 

```suggestion
cdef ITYPE_t[:] _network_simplex(
        ITYPE_t[:] heads,  # IN
        ITYPE_t[:] tails,  # IN
        ITYPE_t[:] capacities,  # IN
        ITYPE_t[:] demand,  # IN
        ITYPE_t[:] cost,  # IN
        ITYPE_t n_verts,  # IN
        ITYPE_t[:] row,  # IN/OUT
        ITYPE_t[:] col,  # IN/OUT
        ITYPE_t[:] flow_data  # IN/OUT
        ):
    cdef:
        ITYPE_t n_edges = capacities.shape[0]
        ITYPE_t idx, faux_inf, capacities_sum, cost_sum
        ITYPE_t j, s, t, i, tmp, p, q, n_non_zero_edges, v
        ITYPE_t flow_cost, flow_value
        bint prev_ret_value = False

        ITYPE_t[:] edge_sources = np.empty(n_edges + n_verts + 1, dtype=ITYPE)
        ITYPE_t[:] edge_targets = np.empty(n_edges + n_verts + 1, dtype=ITYPE)
        ITYPE_t[:] edge_capacities = np.empty(n_edges + n_verts + 1, dtype=ITYPE)
        ITYPE_t[:] edge_weights = np.empty(n_edges + n_verts + 1, dtype=ITYPE)

        ITYPE_t[:] edge_flow = np.zeros(n_edges + n_verts, dtype=ITYPE)

        ITYPE_t[:] vertex_potentials = np.empty(n_verts + 1, dtype=ITYPE)
        ITYPE_t[:] parent = np.empty(n_verts + 1, dtype=ITYPE)
        ITYPE_t[:] parent_edge = np.empty(n_edges + n_verts, dtype=ITYPE)
        ITYPE_t[:] subtree_size = np.empty(n_verts + 1, dtype=ITYPE)
        ITYPE_t[:] next_vertex_dft = np.empty(n_verts + 1, dtype=ITYPE)
        ITYPE_t[:] prev_vertex_dft = np.empty(n_verts + 1, dtype=ITYPE)
        ITYPE_t[:] last_descendent_dft = np.empty(n_verts + 1, dtype=ITYPE)
    
        ITYPE_t[:] Wn = np.empty(n_edges + 1, dtype=ITYPE)
        ITYPE_t[:] We = np.empty(n_edges + 1, dtype=ITYPE)
        ITYPE_t[:] WnR = np.empty(n_verts, dtype=ITYPE)
        ITYPE_t[:] WeR = np.empty(n_edges, dtype=ITYPE)
        ITYPE_t[:] ancestors = np.empty(n_verts + 1, dtype=ITYPE)
        ITYPE_t[:] subtree = np.empty(2 * n_verts, dtype=ITYPE)

        # TODO: use individual variables here
        ITYPE_t[:] local_vars = np.empty(4, dtype=ITYPE)
        ITYPE_t[:] result = np.empty(4, dtype=ITYPE)

    with nogil:
        idx = 0
        n_non_zero_edges = 0
        for e in range(n_edges):
            if capacities[e] != 0:
                n_non_zero_edges += 1
                edge_sources[idx] = tails[e] + 1
                edge_targets[idx] = heads[e] + 1
                idx += 1
        for v in range(n_verts):
            if demand[v] > 0:
                edge_sources[idx] = 0
                edge_targets[idx] = v + 1
            else:
                edge_sources[idx] = v + 1
                edge_targets[idx] = 0
            idx += 1

        capacities_sum, cost_sum = 0, 0
        for e in range(n_edges):
            if capacities[e] < 1 << 31 - 1:
                capacities_sum += capacities[e]
            cost_sum += abs(cost[e])
        faux_inf = max(cost_sum, capacities_sum)
        for v in range(n_verts):
            faux_inf = max(faux_inf, demand[v])
        faux_inf = 3*faux_inf or 1
        idx = 0
        for e in range(n_edges):
            if capacities[e] != 0:
                edge_capacities[idx] = capacities[e]
                edge_weights[idx] = cost[e]
                idx += 1
        for v in range(n_verts):
            edge_capacities[idx] = faux_inf
            edge_weights[idx] = faux_inf
            idx += 1

        _initialize_spanning_tree(n_verts, n_non_zero_edges, faux_inf,
                                demand, edge_flow, vertex_potentials,
                                parent, parent_edge, subtree_size,
                                next_vertex_dft, prev_vertex_dft,
                                last_descendent_dft)

        prev_ret_value = _find_entering_edges(n_non_zero_edges, edge_weights,
                                            vertex_potentials, edge_flow,
                                            edge_sources, edge_targets,
                                            prev_ret_value, local_vars,
                                            result)
        itr = 0
        while prev_ret_value:
            Wne_len = np.empty(2, dtype=ITYPE)
            i, p, q = result[0], result[1], result[2]
            _find_cycle(i, p, q,
                        subtree_size,
                        parent, parent_edge,
                        n_edges, n_verts,
                        WnR, WeR, Wn, We,
                        Wne_len)
            _find_leaving_edge(Wn, We, Wne_len,
                            edge_sources, edge_targets,
                            edge_capacities, edge_flow,
                            result)
            j = result[0]
            s = result[1]
            t = result[2]
            _augment_flow(Wn, We, Wne_len,
                        _residual_capacity(j, s, edge_sources,
                                            edge_capacities,
                                            edge_flow),
                        edge_sources, edge_flow)
            # Do nothing more if the entering edge is the same as the leaving edge.
            if i != j:
                if parent[t] != s:
                    # Ensure that s is the parent of t.
                    tmp = t
                    t = s
                    s = tmp
                if _index(We, Wne_len[1], i) > _index(We, Wne_len[1], j):
                    # Ensure that q is in the subtree rooted at t.
                    tmp = q
                    q = p
                    p = tmp
                _remove_edge(s, t, subtree_size,
                            prev_vertex_dft,
                            last_descendent_dft,
                            next_vertex_dft,
                            parent, parent_edge,
                            n_verts)
                _make_root(n_verts, q,
                        subtree_size,
                        last_descendent_dft,
                        prev_vertex_dft,
                        next_vertex_dft,
                        parent, parent_edge,
                        ancestors)
                _add_edge(i, p, q,
                        last_descendent_dft,
                        next_vertex_dft,
                        prev_vertex_dft,
                        subtree_size,
                        parent, parent_edge)
                _update_potentials(i, p, q,
                                edge_targets, edge_weights,
                                last_descendent_dft,
                                next_vertex_dft,
                                vertex_potentials, subtree)
            prev_ret_value = _find_entering_edges(n_non_zero_edges, edge_weights,
                                                vertex_potentials, edge_flow,
                                                edge_sources, edge_targets,
                                                prev_ret_value, local_vars,
                                                result)
            itr += 1

        result[3] = 1
        for v in range(n_verts):
            if edge_flow[v + n_non_zero_edges] != 0:
                result[3] = 0
                break

        if result[3] == 1:
            flow_cost = 0
            flow_value = 0
            for i in range(n_edges):
                flow_cost += edge_weights[i]*edge_flow[i]
                flow_value += edge_flow[i]
            idx = 0
            for i in range(n_edges):
                if edge_flow[i] != 0 and edge_capacities[i] != 0:
                    _add_entry(edge_sources[i], edge_targets[i],
                            edge_flow[i], idx, row, col, flow_data)
                    idx += 1
            result[0] = flow_cost
            result[1] = idx
            result[2] = flow_value
        return result
    # end: with nogil
```